File: python/03_text_to_speech.py
"""
Text-to-Speech — Vocametrix API example

Synthesizes speech from text and returns the audio with per-character
timing data (useful for karaoke highlighting or lip-sync animation).

Auth: X-API-Key header
Pattern: synchronous POST — audio returned directly in response

Usage:
    python 03_text_to_speech.py "Hello, this is a test."
    python 03_text_to_speech.py "Bonjour le monde" --locale fr-FR --output speech.wav
"""

import os
import json
import argparse
import requests
from dotenv import load_dotenv

load_dotenv(os.path.join(os.path.dirname(__file__), '..', '.env'))
API_KEY = os.environ['VOCAMETRIX_API_KEY']
BASE_URL = 'https://platform.vocametrix.com'
HEADERS = {'X-API-Key': API_KEY}


def synthesize(text: str, locale: str, voice: str | None = None) -> dict:
    payload = {'text': text, 'locale': locale}
    if voice:
        payload['voiceName'] = voice

    r = requests.post(f'{BASE_URL}/api/text-to-speech', headers=HEADERS, json=payload)
    r.raise_for_status()
    return r.json()


def print_results(result: dict, output_path: str | None):
    print('\n=== Text-to-Speech Results ===')
    print(f"Duration: {result.get('audioDuration', 'N/A')} seconds")

    timing = result.get('charTimings', result.get('characterTimings', []))
    if timing:
        print(f'Character timings: {len(timing)} entries')
        print('First 5:')
        for entry in timing[:5]:
            print(f"  char={entry.get('char', '?')!r:4s}  offset={entry.get('audioOffset', 0) / 1e7:.3f}s")

    audio_b64 = result.get('audioData', result.get('audio', ''))
    if audio_b64 and output_path:
        import base64
        with open(output_path, 'wb') as f:
            f.write(base64.b64decode(audio_b64))
        print(f'\nAudio saved to {output_path}')
    elif audio_b64:
        print(f'\nAudio data returned ({len(audio_b64)} base64 chars). Pass --output to save.')


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Text-to-speech synthesis')
    parser.add_argument('text', help='Text to synthesize')
    parser.add_argument('--locale', default='en-US', help='Language locale')
    parser.add_argument('--voice', help='Voice name (optional)')
    parser.add_argument('--output', help='Save audio to this WAV path')
    args = parser.parse_args()

    result = synthesize(args.text, args.locale, args.voice)
    print_results(result, args.output)

File: javascript/03-text-to-speech.js
/**
 * Text-to-Speech — Vocametrix API example (Node.js)
 *
 * Synthesizes speech and returns audio with per-character timing data.
 * Useful for karaoke highlighting or lip-sync animation.
 *
 * Auth: X-API-Key header
 * Pattern: synchronous POST
 *
 * Usage:
 *   node 03-text-to-speech.js "Hello, this is a test."
 *   node 03-text-to-speech.js "Bonjour" --locale fr-FR --output speech.wav
 */

import fs from 'fs';
import 'dotenv/config';

const API_KEY = process.env.VOCAMETRIX_API_KEY;
const BASE_URL = 'https://platform.vocametrix.com';
const HEADERS = { 'X-API-Key': API_KEY, 'Content-Type': 'application/json' };

async function synthesize(text, locale, voice) {
  const body = { text, language: locale };
  if (voice) body.voice = voice;

  const res = await fetch(`${BASE_URL}/api/text-to-speech`, {
    method: 'POST',
    headers: HEADERS,
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`text-to-speech failed: ${await res.text()}`);
  return res.json();
}

function printResults(result, outputPath) {
  console.log('\n=== Text-to-Speech Results ===');
  console.log(`Duration: ${result.audioDuration ?? 'N/A'} seconds`);

  const timing = result.charTimings ?? result.characterTimings ?? [];
  if (timing.length > 0) {
    console.log(`Character timings: ${timing.length} entries`);
    console.log('First 5:');
    for (const entry of timing.slice(0, 5)) {
      const offset = ((entry.audioOffset ?? 0) / 1e7).toFixed(3);
      console.log(`  char=${JSON.stringify(entry.char ?? '?').padEnd(4)}  offset=${offset}s`);
    }
  }

  const audioB64 = result.audioData ?? result.audio ?? '';
  if (audioB64 && outputPath) {
    fs.writeFileSync(outputPath, Buffer.from(audioB64, 'base64'));
    console.log(`\nAudio saved to ${outputPath}`);
  } else if (audioB64) {
    console.log(`\nAudio data returned (${audioB64.length} base64 chars). Pass --output to save.`);
  }
}

// --- CLI ---
const args = process.argv.slice(2);
if (args.length === 0) {
  console.error('Usage: node 03-text-to-speech.js <text> [--locale en-US] [--voice name] [--output file.wav]');
  process.exit(1);
}
const text = args[0];
const localeIdx = args.indexOf('--locale');
const voiceIdx = args.indexOf('--voice');
const outputIdx = args.indexOf('--output');
const locale = localeIdx !== -1 ? args[localeIdx + 1] : 'en-US';
const voice = voiceIdx !== -1 ? args[voiceIdx + 1] : null;
const output = outputIdx !== -1 ? args[outputIdx + 1] : null;

synthesize(text, locale, voice)
  .then(r => printResults(r, output))
  .catch(err => { console.error(err.message); process.exit(1); });
